qos.go

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2015-2025 MinIO, Inc.
+//
+// This file is part of MinIO Object Storage stack
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+//
+
+package madmin
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"net/url"
+
+	"gopkg.in/yaml.v2"
+)
+
+// QOSConfigVersionCurrent is the current version of the QoS configuration.
+const QOSConfigVersionCurrent = "v1"
+
+// QOSConfig represents the QoS configuration with a list of rules.
+type QOSConfig struct {
+	Version string    `yaml:"version"` // Version of the QoS configuration format
+	Rules   []QOSRule `yaml:"rules"`
+}
+
+// QOSRule represents a single QoS rule at the bucket prefix or bucket wildcard level.
+type QOSRule struct {
+	ID           string          `yaml:"id"`
+	Label        string          `yaml:"label,omitempty"`
+	BucketPrefix string          `yaml:"bucketPrefix"` // includes prefix if any and represents simple wildcard for bucket, prefix
+	Requests     []RequestLimits `yaml:"requests,omitempty" json:"requests,omitempty" mapstructure:"requests"`
+}
+
+// RequestLimits - define Request Limits
+type RequestLimits struct {
+	API      string `yaml:"api" json:"api" mapstructure:"api"`
+	APILimit `yaml:",inline" mapstructure:",squash"`
+}
+
+// APILimit - define API Limit
+type APILimit struct {
+	Rate  int64  `yaml:"rate" json:"rate" mapstructure:"rate"`    // requests per second for (rps), concurrent requests for (concurrency)
+	Burst int64  `yaml:"burst" json:"burst" mapstructure:"burst"` // burst size for the limiter
+	Limit string `yaml:"limit" json:"type" mapstructure:"limit"`  // "concurrency" or "rps" limiter
```

Comment:
maybe turns this into types ? anyone that's using the api would have to read the docs for this to make sense